### ISSUE TITLE
Azure: When using Managed DNS Hive fails to find a SOA record

### ISSUE THREAD

**Issue by m1kola** [openshift/hive#1720]
When using Managed DNS on Azure Hive fails to find a SOA record.

## Steps to reproduce

1. Create a DNS zone (I use `hive.example.com` for example)
1. Create/update `HiveConfig` as [per docs](https://github.com/openshift/hive/blob/bfb69aec0a7206918012bb21ed25cae098401b59/docs/using-hive.md#managed-dns-1) so that `hive.example.com` is included in `managedDomains`.
1. Create a cluster
    1. In the cluster install config set `baseDomain: cluster.hive.example.com`
    1. In `ClusterDeployment` the spec should include the following:

        ```yaml
        spec:
          manageDNS: true
          baseDomain: cluster.hive.example.com
        # ...
        ```
1. Run `oc logs -f deployment.apps/hive-controllers` to watch for logs.

You should see in logs that the controller creates a child zone `cluster.hive.example.com`, but then fails to find a SOA record for that zone:

```
level=info msg="reconciling dns zone" controller=dnszone dnsZone=example-cluster/cluster-zone reconcileID=f7jzgsr2
level=debug msg="DNSZone is not involved in a relocate" controller=dnszone dnsZone=example-cluster/cluster-zone reconcileID=f7jzgsr2
level=info msg="Syncing DNS Zone" controller=dnszone currentGeneration=1 delta=0s dnsZone=example-cluster/cluster-zone lastSyncGeneration=1 reconcileID=f7jzgsr2
level=debug msg="Retrieving current state" controller=dnszone
level=debug msg="Fetching managed zone by zone name" controller=dnszone dnsZone=example-cluster/cluster-zone reconcileID=f7jzgsr2 zone=cluster.hive.example.com
level=debug msg="Found managed zone" controller=dnszone dnsZone=example-cluster/cluster-zone reconcileID=f7jzgsr2 zone=cluster.hive.example.com
level=info msg="Existing hosted zone found. Syncing with DNSZone resource" controller=dnszone
level=debug msg="found managed zone name servers" controller=dnszone dnsZone=example-cluster/cluster-zone nameservers="&[ns1-03.azure-dns.com. ns2-03.azure-dns.net. ns3-03.azure-dns.org. ns4-03.azure-dns.info.]" reconcileID=f7jzgsr2 zone=cluster.hive.example.com
level=info msg="looking up domain SOA record" controller=dnszone servers="[172.30.0.10:53]"
level=info msg="SOA query duration: 25.958524ms" server="172.30.0.10:53"
level=info msg="no answer for SOA record returned" controller=dnszone server="172.30.0.10:53"
level=info msg="SOA record for DNS zone not available" controller=dnszone
level=debug msg="Updating DNSZone status" controller=dnszone
level=info msg="reconcile complete" controller=dnszone dnsZone=example-cluster/cluster-zone elapsedMillis=702 elapsedMillisGT=0 outcome=unspecified reconcileID=f7jzgsr2
```

You should also be able to find `cluster.hive.example.com` resource on Azure portal or using az CLI.

## Potential cause

The controller doesn't seem to create NS records in the parent zone `hive.example.com` to allow delegation of queries about `cluster.hive.example.com` to the right name server.

I believe this is why hive controller is not able to find a SOA record.

**Comment by staebler** [comment]
> level=info msg="Existing hosted zone found. Syncing with DNSZone resource" controller=dnszone

This is the problematic part, if I recall correctly from my outdated experience. Hive will only do the NS records if it is creating the zone for the cluster's base domain. It that zone already existed, then Hive must assume that the zone is already functional.

**Comment by m1kola** [comment]
> This is the problematic part, if I recall correctly from my outdated experience. Hive will only do the NS records if it is creating the zone for the cluster's base domain. It that zone already existed, then Hive must assume that the zone is already functional.

@staebler Hive did create a new zone `cluster.hive.example.com` as it did not exist prior to creation of `ClusterDeployment`, but did not create `NS` records. The logs I provided above are from Hive controller which is stuck in a reconciliation loop. Unfortunately I did not capture logs from the first attempt to reconcile the DNS zone (when it was actually created by the hive controller).

I had a very quick look at the Hive codebase and I do not see any code which sets NS records on DNS zone creation at all: I only see zone creation itself:

https://github.com/openshift/hive/blob/bfb69aec0a7206918012bb21ed25cae098401b59/pkg/controller/dnszone/azureactuator.go#L69-L75

And implementation of `CreateOrUpdateZone`:

https://github.com/openshift/hive/blob/bfb69aec0a7206918012bb21ed25cae098401b59/pkg/azureclient/client.go#L68-L75

Nothing about NS records here or anywhere near apart from getting nameservers in the actuator here:

https://github.com/openshift/hive/blob/bfb69aec0a7206918012bb21ed25cae098401b59/pkg/controller/dnszone/azureactuator.go#L147-L157


I have a feeling that there was an assumption that `CreateOrUpdateZone` API call will create a delegated DNS zone in Azure.
